### PR TITLE
Fixing issues in Governance Connectors related to secondary user stores

### PR DESCRIPTION
**Pull request by ayshsandu**
IDENTITY-5265
IDENTITY-5333
IDENTITY-5338
IDENTITY-5308
IDENTITY-5291
IDENTITY-5262

### CODE REVIEW

File: components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/handler/AdminForcedPasswordResetHandler.java [hunk 1]
```diff
@@ -54,13 +54,16 @@ public void handleEvent(Event event) throws IdentityEventException {
                 .EventProperty.USER_STORE_MANAGER);
         User user = getUser(eventProperties, userStoreManager);
```
Comment (darshanasbg) [inline components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/handler/AdminForcedPasswordResetHandler.java]: It's better load user only for matching event types.

File: components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/handler/UserEmailVerificationHandler.java [hunk 2]
```diff
@@ -68,6 +68,9 @@ public void handleEvent(Event event) throws IdentityEventException {
         UserStoreManager userStoreManager = (UserStoreManager) eventProperties.get(IdentityEventConstants.EventProperty.USER_STORE_MANAGER);
         User user = getUser(eventProperties, userStoreManager);
 
+        if(user == null){
```
Comment (darshanasbg) [inline components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/handler/UserEmailVerificationHandler.java]: How does it get `user` null here?

File: components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/handler/UserEmailVerificationHandler.java [hunk 3]
```diff
@@ -248,6 +250,14 @@ protected void triggerNotification(User user, String type, String code, Property
     protected User getUser(Map eventProperties, UserStoreManager userStoreManager){
 
         String userName = (String) eventProperties.get(IdentityEventConstants.EventProperty.USER_NAME);
+        try {
+            if(!userStoreManager.isExistingUser(userName)){
```
Comment (darshanasbg) [inline components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/handler/UserEmailVerificationHandler.java]: What is the scenario, a non existing user triggering this flow and get to this point?

As I see, this handler registered to PRE_ADD_USER, POST_ADD_USER events, and those are getting triggered only for the user store that user exists.

File: components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/handler/UserSelfRegistrationHandler.java [hunk 4]
```diff
@@ -67,6 +67,14 @@ public void handleEvent(Event event) throws IdentityEventException {
         String userName = (String) eventProperties.get(IdentityEventConstants.EventProperty.USER_NAME);
         UserStoreManager userStoreManager = (UserStoreManager) eventProperties.get(IdentityEventConstants.EventProperty.USER_STORE_MANAGER);
 
+        try {
+            if(!userStoreManager.isExistingUser(userName)){
+               return;
+            }
+        } catch (UserStoreException e) {
+            log.error("Error while checking user : " + userName +" in userStore.", e);
```
Comment (darshanasbg) [inline components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/handler/UserSelfRegistrationHandler.java]: Fix formatting issues

File: components/org.wso2.carbon.identity.account.suspension.notification.task/src/main/java/org/wso2/carbon/identity/account/suspension/notification/task/handler/AccountSuspensionNotificationHandler.java [hunk 5]
```diff
@@ -56,10 +56,11 @@ public void handleEvent(Event event) throws IdentityEventException {
             String tenantDomain = (String) eventProperties.get(IdentityEventConstants.EventProperty.TENANT_DOMAIN);
             UserStoreManager userStoreManager = (UserStoreManager) eventProperties
                     .get(IdentityEventConstants.EventProperty.USER_STORE_MANAGER);
-
             try {
-                userStoreManager.setUserClaimValue(userName, IdentityMgtConstants.LAST_LOGIN_TIME,
-                        Long.toString(System.currentTimeMillis()), null);
+                if(userStoreManager.isExistingUser(userName)){
```
Comment (ayshsandu) [inline components/org.wso2.carbon.identity.account.suspension.notification.task/src/main/java/org/wso2/carbon/identity/account/suspension/notification/task/handler/AccountSuspensionNotificationHandler.java]: Added check for authenticated users

Comment (ayshsandu) [inline components/org.wso2.carbon.identity.account.suspension.notification.task/src/main/java/org/wso2/carbon/identity/account/suspension/notification/task/handler/AccountSuspensionNotificationHandler.java]: Moved the logic to listener level

File: components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/handler/AccountConfirmationValidationHandler.java [hunk 6]
```diff
@@ -75,6 +75,9 @@ public void handleEvent(Event event) throws IdentityEventException {
                 }
             boolean isAccountLocked = true ;
             try {
+                if(!userStoreManager.isExistingUser(userName)){
```
Comment (ayshsandu) [inline components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/handler/AccountConfirmationValidationHandler.java]: Still in discussion? Has to be done with care.

Comment (ayshsandu) [inline components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/handler/AccountConfirmationValidationHandler.java]: refactored.

File: components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/handler/AdminForcedPasswordResetHandler.java [hunk 7]
```diff
@@ -54,13 +54,16 @@ public void handleEvent(Event event) throws IdentityEventException {
...
 
-        if (IdentityEventConstants.Event.PRE_SET_USER_CLAIMS.equals(eventName)) {
-            handleClaimUpdate(eventProperties, userStoreManager, user);
-        }
+        if(user != null){
```
Comment (ayshsandu) [inline components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/handler/AdminForcedPasswordResetHandler.java]: refactored.

File: components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/handler/UserEmailVerificationHandler.java (hunk 2 above)
Comment (ayshsandu) [inline components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/handler/UserEmailVerificationHandler.java]: Changed when logic moved to listener level

File: components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/handler/UserEmailVerificationHandler.java (hunk 3 above)
Comment (ayshsandu) [inline components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/handler/UserEmailVerificationHandler.java]: Fixed as isUserExistsInDomain moved to listener level

File: components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/handler/UserSelfRegistrationHandler.java (hunk 4 above)
Comment (ayshsandu) [inline components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/handler/UserSelfRegistrationHandler.java]: Done

File: components/org.wso2.carbon.identity.governance/src/main/java/org/wso2/carbon/identity/governance/listener/IdentityMgtEventListener.java [hunk 8]
```diff
@@ -77,6 +80,13 @@ public boolean doPreAuthenticate(String userName, Object credential,
         if (log.isDebugEnabled()) {
             log.debug("Pre authenticator is called in IdentityMgtEventListener");
         }
+        if (!isUserExistsInDomain(userStoreManager, userName)) {
+            if (log.isDebugEnabled()) {
+                log.debug("IdentityMgtEventListener returns since user: " + userName + " not available is current " +
```
Comment (darshanasbg) [inline components/org.wso2.carbon.identity.governance/src/main/java/org/wso2/carbon/identity/governance/listener/IdentityMgtEventListener.java]: Better to log userstore domain here

File: components/org.wso2.carbon.identity.governance/src/main/java/org/wso2/carbon/identity/governance/listener/IdentityMgtEventListener.java [hunk 9]
```diff
@@ -94,6 +104,13 @@ public boolean doPostAuthenticate(String userName, boolean authenticated, UserSt
         if (log.isDebugEnabled()) {
             log.debug("post authenticator is called in IdentityMgtEventListener");
         }
+        if (!isUserExistsInDomain(userStoreManager, userName, authenticated)){
+            if (log.isDebugEnabled()) {
+                log.debug("IdentityMgtEventListener returns since user: " + userName + " not available is current " +
```
Comment (darshanasbg) [inline components/org.wso2.carbon.identity.governance/src/main/java/org/wso2/carbon/identity/governance/listener/IdentityMgtEventListener.java]: Log userstore name

File: components/org.wso2.carbon.identity.governance/src/main/java/org/wso2/carbon/identity/governance/listener/IdentityMgtEventListener.java [hunk 10]
```diff
@@ -641,4 +658,32 @@ public boolean doPostGetUserClaimValues(String userName, String[] claims, String
...
+
+    private boolean isUserExistsInDomain(UserStoreManager userStoreManager, String userName) throws UserStoreException {
+        boolean isExists = false;
+        if (userStoreManager.isExistingUser(userName)) {
+            isExists = true;
+        }
+        return isExists;
+    }
+
+    private boolean isUserExistsInDomain(UserStoreManager userStoreManager, String userName, boolean authenticated) throws UserStoreException {
```
Comment (darshanasbg) [inline components/org.wso2.carbon.identity.governance/src/main/java/org/wso2/carbon/identity/governance/listener/IdentityMgtEventListener.java]: Fix formatting issues

File: components/org.wso2.carbon.identity.governance/src/main/java/org/wso2/carbon/identity/governance/listener/IdentityMgtEventListener.java (hunk 9 above)
Comment (ayshsandu) [inline components/org.wso2.carbon.identity.governance/src/main/java/org/wso2/carbon/identity/governance/listener/IdentityMgtEventListener.java]: User Store domain appended to log

File: components/org.wso2.carbon.identity.governance/src/main/java/org/wso2/carbon/identity/governance/listener/IdentityMgtEventListener.java (hunk 8 above)
Comment (ayshsandu) [inline components/org.wso2.carbon.identity.governance/src/main/java/org/wso2/carbon/identity/governance/listener/IdentityMgtEventListener.java]: User Store domain appended to log

**Pull request merged**